### ISSUE TITLE
Bug for SelectableText. If text contains chinese words and '\r', the selectbox will deviation  

### ISSUE THREAD

**Issue by droplet-js** [flutter/flutter#65905]
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the files as follows: ... 
```dart
import 'package:flutter/material.dart';

class App extends StatefulWidget {
  const App({Key key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _AppState();
  }
}

class _AppState extends State<App> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('SelectableText'),
        ),
        body: SelectableText(
          '汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n'
        ),
      ),
    );
  }
}
```
3. select all text

Expected results           | Actual results
:-------------------------:|:-------------------------:
![93311794-05104c80-f839-11ea-8a16-4f8ccc38e7b3](https://user-images.githubusercontent.com/8845083/94395528-7e416500-0192-11eb-8381-3350eca19205.jpg) | ![Screenshot_20200916_102224_io github v7lin exampl](https://user-images.githubusercontent.com/8845083/93285147-c366ae00-f806-11ea-810f-ec99d4a7e7ea.jpg)

<details>
  <summary>Logs</summary>

```
```

```
Analyzing bug...                                                     
No issues found! (ran in 4.6s)
```

```
[✓] Flutter (Channel stable, 1.20.3, on Mac OS X 10.15.5 19F101, locale zh-Hans-CN)
    • Flutter version 1.20.3 at /Users/v7lin/flutter
    • Framework revision 216dee60c0 (2 周前), 2020-09-01 12:24:47 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/v7lin/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/v7lin/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (1 available)
    • VCE AL00 (mobile) • JPFDU19523006521 • android-arm64 • Android 10 (API 29)

• No issues found!
```

</details>

**Comment by amanv8060** [comment]
Please also mention what is expected

**Comment by amanv8060** [comment]
And Also this isnt a bug due to carriage return , but due to chinese characters ,
`'汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n'` this would also give you the same result

**Comment by darshankawar** [comment]
@v7lin 
Please try to use chinese equivalent of `\r` and see if you get the same result. Also not sure what should be the expected outcome of the usage.
Thanks.

**Comment by v7lin** [comment]
i try to replace '\r' to '\u3000'，it work well

```dart
import 'package:flutter/material.dart';

class App extends StatefulWidget {
  const App({Key key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _AppState();
  }
}

class _AppState extends State<App> {

  static const String sbcSpace = '\u3000';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('SelectableText'),
        ),
        body: SelectableText(
          '汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n'.replaceAll('\r', sbcSpace)
        ),
      ),
    );
  }
}
```

![Screenshot_20200916_162205_io github v7lin exampl](https://user-images.githubusercontent.com/8845083/93311794-05104c80-f839-11ea-8a16-4f8ccc38e7b3.jpg)

**Comment by darshankawar** [comment]
@v7lin is it working as intended now ?

**Comment by v7lin** [comment]
I dont agree. '\r' dont have chinese equivalent character.

**Comment by gtyhn** [comment]
No, it should not be closed, the issue still exists on 1.20.4, and in Chinese characters, if there are multiple line breaks, it can cause the WidgetSpan position to move down

**Comment by gtyhn** [comment]
when using non-English characters, the position in the text where WidgetSpan is used is affected.

![bug](https://user-images.githubusercontent.com/8544251/94272039-16620300-ff75-11ea-9ed7-0a5a561c1327.png)

**Comment by gtyhn** [comment]
@darshankawar

**Comment by amanv8060** [comment]
https://github.com/flutter/flutter/issues/65905#issuecomment-693168413 , i mentioned above , i guess it was ignored because i was not the author of issue , anyways cc @darshankawar

**Comment by darshankawar** [comment]
@v7lin Please provide what should be the expected result when you select the entire text. 
@gtyhn same ask is from you as well. Please provide what should be the expected result along with `flutter doctor -v` and minimal code sample.
Thanks

**Comment by v7lin** [comment]
i have mention what is expected

**Comment by darshankawar** [comment]
Issue replicable on latest stable (1.20.4). After selecting all text, the selectable text area doesn't cover the actual text entirely till the end of the list.

![65905](https://user-images.githubusercontent.com/67046386/94436750-71386c00-01ba-11eb-8542-f798c75706f6.png)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.20.4, on Mac OS X 10.15.4 19E2269, locale en-IN)
    • Flutter version 1.20.4 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision fba99f6cf9 (13 days ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (1 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm • Android 8.1.0 (API
      27)

• No issues found!

```

**Comment by gtyhn** [comment]
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('test'),
      ),
      body: SingleChildScrollView(
        child: _buildRich(),
      ),
    );
  }

  final _text = '汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n';
  final _text2 = '汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n';

  Widget _buildRich() {
    return RichText(
      text: TextSpan(
        style: TextStyle(
          fontSize: 16,
          color: Colors.black,
        ),
        children: [
          TextSpan(text: _text),
          _buildWidget(),
          TextSpan(text: _text2),
          _buildWidget(),
          TextSpan(text: _text2),
        ],
      ),
    );
  }

  InlineSpan _buildWidget() {
    return WidgetSpan(
      child: Container(
        width: double.infinity,
        height: 100,
        decoration: BoxDecoration(
          color: Colors.black12,
          border: Border.all(
            color: Colors.red,
          ),
        ),
        child: Text(
          'WidgetSpan -> Container',
          textAlign: TextAlign.center,
        ),
      ),
    );
  }
}
```

<details>
<summary>flutter doctor -v</summary>

```
[√] Flutter (Channel stable, 1.20.4, on Microsoft Windows [Version 10.0.18363.1082], locale zh-CN)
    • Flutter version 1.20.4 at C:\flutter
    • Framework revision fba99f6cf9 (2 weeks ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at D:\android\sdk\android-sdk-windows
    • Platform android-30, build-tools 30.0.0
    • Java binary at: D:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    X Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed       
      instructions.

[√] Android Studio (version 4.0)
    • Android Studio at D:\Program Files\Android\Android Studio
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (1 available)
    • LIO AN00 (mobile) • 127.0.0.1:62025 • android-x86 • Android 7.1.2 (API 25)
```

</details>

![bug](https://user-images.githubusercontent.com/8544251/94509810-c4a4cb80-0247-11eb-83db-f9535073ee9a.png)

@darshankawar 
so this problem still exists

**Comment by gtyhn** [comment]
@darshankawar 

Is there a plan to solve this problem? The inclusion of a line break in a Chinese word causes WidgetSpan to be incorrectly positioned.

**Comment by v7lin** [comment]
it fixed on latest stable (1.22.1)

**Comment by maheshmnj** [comment]
@gtyhn @droplet-js I tried to replicate the issue on the latest maser channel and I am unable to reproduce the issue. So I believe the issue is resolved. I tried both of your code samples and this is the output I found 

code sample 1             |  code sample 2
:-------------------------:|:-------------------------:
![Simulator Screen Shot - iPhone 12 Pro - 2022-11-25 at 16 35 12](https://user-images.githubusercontent.com/31410839/204057672-a9cd8c59-8496-4c06-81ec-5a6fb6ce5b2f.png) | ![Simulator Screen Shot - iPhone 12 Pro - 2022-11-25 at 16 39 03](https://user-images.githubusercontent.com/31410839/204057685-7198f48f-c3f9-430b-be17-3c2ad76d93d9.png)

<details>
<summary>code sample 1</summary>

```dart
import 'package:flutter/material.dart';

void main(List<String> args) {
  runApp(App());
}

class App extends StatefulWidget {
  const App({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _AppState();
  }
}

class _AppState extends State<App> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('SelectableText'),
        ),
        body: const SelectableText(
          '汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n汉字\r\n\r\n'
        ),
      ),
    );
  }
}
```
</details> 

<details>
<summary>code sample 2</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('test'),
      ),
      body: SingleChildScrollView(
        child: _buildRich(),
      ),
    );
  }

  final _text = '汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n';
  final _text2 = '汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n汉字\n\n';

  Widget _buildRich() {
    return RichText(
      text: TextSpan(
        style: TextStyle(
          fontSize: 16,
          color: Colors.black,
        ),
        children: [
          TextSpan(text: _text),
          _buildWidget(),
          TextSpan(text: _text2),
          _buildWidget(),
          TextSpan(text: _text2),
        ],
      ),
    );
  }

  InlineSpan _buildWidget() {
    return WidgetSpan(
      child: Container(
        width: double.infinity,
        height: 100,
        decoration: BoxDecoration(
          color: Colors.black12,
          border: Border.all(
            color: Colors.red,
          ),
        ),
        child: Text(
          'WidgetSpan -> Container',
          textAlign: TextAlign.center,
        ),
      ),
    );
  }
}
```
</details> 


<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel master, 3.6.0-6.0.pre.33, on macOS 12.6 21G115 darwin-arm64,
    locale en-IN)
    • Flutter version 3.6.0-6.0.pre.33 on channel master at
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a9858ec524 (72 minutes ago), 2022-11-21 10:58:11 -0500
    • Engine revision 46a6b54295
    • Dart version 2.19.0 (build 2.19.0-406.0.dev)
    • DevTools version 2.19.0
    • If those were intentional, you can disregard the above warnings; however it is
      recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.53.20221101

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 026D5789-9E78-4AD5-B1B2-3F8D4E7F65E4 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64
      • macOS 12.6 21G115 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript
      • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
```
[✓] Flutter (Channel stable, 3.3.9, on macOS 12.6 21G115 darwin-arm, locale en-IN)
    • Flutter version 3.3.9 on channel stable at /Users/mahesh/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b8f7f1f986 (24 hours ago), 2022-11-23 06:43:51 +0900
    • Engine revision 8f2221fbef
    • Dart version 2.18.5
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.53.20221101

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 026D5789-9E78-4AD5-B1B2-3F8D4E7F65E4 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.6 21G115 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>


I am going to close this issue as fixed, In case you still encounter the issue, feel free to write in the comments and we will reopen it right away.

Thank you.
